aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -67,77 +137,64 @@ namespace detail {
     AT_ASSERT(device >= 0 && device < num_gpus);
   }
 
-  CUDAStreamInternals* CUDAStream_getDefaultStreamOnDevice(int64_t device) {
-    initCUDAStreamsOnce();
-    check_gpu(device);
-    return &default_streams[device];
-  }
-  CUDAStreamInternals* CUDAStream_getDefaultStream() {
-    return CUDAStream_getDefaultStreamOnDevice(current_device());
-  }
-
-  // Creates (and retains) and new cuda stream
-  CUDAStreamInternals* CUDAStream_createAndRetainWithOptions(int32_t flags, int32_t priority) {
-    CUDAStreamInternals* internals = (CUDAStreamInternals*) malloc(sizeof(CUDAStreamInternals));
-    internals->is_destructible = true;
-    internals->refcount = 1;
-    internals->device = current_device();
-    #ifndef __HIP_PLATFORM_HCC__
-      AT_CUDA_CHECK(cudaStreamCreateWithPriority(&internals->stream, flags, priority));
-    #else 
-      AT_CUDA_CHECK(cudaStreamCreateWithFlags(&internals->stream, flags));
-    #endif // __HIP_PLATFORM_HCC__
-    return internals;
+  // Helper to determine the index of the stream to return
+  // Streams are returned round-robin, and the counter is "kept" between
+  // 0 and STREAMS_PER_POOl. 
+  // Note: it is possible the counter will grow beyond STREAMS_PER_POOL 
+  // temporarily, but it will eventually return to the desired range.
+  static int get_idx(std::atomic<int> &counter) {
+    int raw_idx = counter++;
+    int modded = raw_idx % STREAMS_PER_POOL;
+    if (raw_idx >= STREAMS_PER_POOL && modded == 0) {
+      counter -= STREAMS_PER_POOL;
+    }
```

Comment:
This is confusing. You're claiming to generate numbers in the `[0, STREAMS_PER_POOL]` interval, but it really looks like you're doing `[1, STREAMS_PER_POOL]`. Apart from that it seems simpler to just do `return (raw_idx % STREAMS_PER_POOL) + 1`

Reply:
Yes, use an unsigned integer type so that it has well-defined wraparound overflow, and just don't bother trying to force it back in the range. Because `STREAMS_PER_POOL` is a power of two, you won't even have imbalance when you get to the wraparound point. (Actually, it looks like signed integers in std::atomic are defined to uses two's complement, so overflow there is defined too.)